src/engine.ts

```diff
@@ -661,6 +741,66 @@ class StyleElementController extends NodeController<HTMLStyleElement> {
   }
 }
 
+class ShadowRootController extends NodeController<ShadowRoot> {
+  private controller: AbortController | null = null;
+  private mo: MutationObserver;
+  private host: HTMLElement;
+
+  constructor(node: ShadowRoot, mo: MutationObserver) {
+    super(node);
+    this.mo = mo;
+    this.host = node.host as HTMLElement;
+  }
+
+  connected(): void {
+    this.mo?.observe(this.node, {
+      childList: true,
+      subtree: true,
+      attributes: true,
+      attributeOldValue: true,
+    });
+  }
+
+  updated() {
+    this.processCssRules();
+  }
+
+  mutated() {
+    this.processCssRules();
+  }
+
+  disconnected(): void {
+    this.controller?.abort();
+    this.controller = null;
+    this.mo?.disconnect();
+  }
+
+  private processCssRules(): void {
+    let setProp = false;
+    for (const adoptedStyleSheet of this.node.adoptedStyleSheets) {
+      for (const rule of adoptedStyleSheet.cssRules) {
+        if (rule instanceof CSSStyleRule) {
+          const value = rule.style.getPropertyValue(CUSTOM_PROPERTY_TYPE).trim();
+          if (value) {
+            const selector: string = rule.selectorText.replace(':host', this.host.localName);
```

Comment:
I still think I would prefer to have this moved into a separate PR. Without these changes, you should be able to use container queries in a shadow DOM, just without supporting the shadow DOM selectors.

Then we can work on finalizing support for those selectors

Reply:
Sorry for my confusion on this part, I didn't realize you were wanting to move this out completely. 

Because we were discussing making this PR the base impl to enable `container-type` functionality from within the shadow DOM, I viewed this as being part of the impl, as it would not be uncommon for developers to want to apply `container-type` to the `host` from within the shadow DOM via the `:host` selector. 
- This is a minimum impl to support `container-type` on the `:host` selector from within the shadow DOM, re-process the rules on mutation events incase attributes are changed on the host, and nothing more.

```css 
:host { container-type: inline-size; }
``` 
or  
```css
:host[attribute] { container-type: inline-size; }
```

Do you still want me to move this out?

Reply:
Taking a look now...

Reply:
@devknoll let me know any thoughts or concerns when you have time to circle back to this.

Reply:
@devknoll after taking a closer look at this, maybe this is violating `CSS Scoping` and it is ok for the `:host` selector to not support defining a `container-type`?

Reply:
Stashed this impl in case it's useful in the future. Went ahead and removed it per your original request.